OvmfPkg/Library/PlatformInitLib/MemDetect.c

```diff
@@ -898,6 +898,69 @@ PlatformScanHostProvided64BitPciMmioEnd (
   return EFI_NOT_FOUND;
 }
 
+VOID
+EFIAPI
+Switch4Level (
+  VOID
+  );
+
+STATIC
+VOID
+PlatformSetupPagingLevel (
+  IN OUT EFI_HOB_PLATFORM_INFO  *PlatformInfoHob
+  )
+{
+ #ifdef MDE_CPU_X64
+  UINT32      PagingLevel;
+  EFI_STATUS  Status;
+  IA32_CR4    Cr4;
+
+  Cr4.UintN = AsmReadCr4 ();
+  if (!Cr4.Bits.LA57) {
+    DEBUG ((DEBUG_INFO, "%a: using 4-level paging (la57 not supported)\n", __func__));
+    return;
+  }
+
+  Status = QemuFwCfgParseUint32 (
+             "opt/org.tianocode/PagingLevel",
+             FALSE,
+             &PagingLevel
+             );
+  switch (Status) {
+    case EFI_NOT_FOUND:
+      if (PlatformInfoHob->FirstNonAddress < (1ll << 40)) {
+        //
+        // If the highest address actually used is below 1TB switch back into
+        // 4-level paging mode for better compatibility with older guests.
+        //
+        DEBUG ((DEBUG_INFO, "%a: using 4-level paging (default for small guest)\n", __func__));
+        PagingLevel = 4;
+      } else {
+        DEBUG ((DEBUG_INFO, "%a: using 5-level paging (default for large guest)\n", __func__));
+        PagingLevel = 5;
+      }
+
+      break;
+    case EFI_SUCCESS:
+      if ((PagingLevel != 4) && (PagingLevel != 5)) {
+        DEBUG ((DEBUG_INFO, "%a: invalid paging level in fw_cfg: %d\n", __func__, PagingLevel));
+        return;
+      }
+
+      DEBUG ((DEBUG_INFO, "%a: using %d-level paging (fw_cfg override)\n", __func__, PagingLevel));
+      break;
+    default:
+      DEBUG ((DEBUG_WARN, "%a: QemuFwCfgParseUint32: %r\n", __func__, Status));
+      return;
+  }
+
+  if (PagingLevel == 4) {
+    Switch4Level ();

```

Comment:
This is only for BSP, what about AP?

Reply:
PlatformPei runs before AP bringup.
Current edk2 will simply propagate BSP configuration to APs, so everything should work just fine.